test-integ/catalogv2/traffic_permissions_test.go

```diff
@@ -0,0 +1,459 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+package catalogv2
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pbauth "github.com/hashicorp/consul/proto-public/pbauth/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"github.com/hashicorp/consul/sdk/testutil/retry"
+	"github.com/hashicorp/consul/test-integ/topoutil"
+	libassert "github.com/hashicorp/consul/test/integration/consul-container/libs/assert"
+	"github.com/hashicorp/consul/test/integration/consul-container/libs/utils"
+	"github.com/hashicorp/consul/testing/deployer/sprawl/sprawltest"
+	"github.com/hashicorp/consul/testing/deployer/topology"
+)
+
+type testCase struct {
+	permissions []*permission
+	result      []*testResult
+}
+
+type permission struct {
+	allow                bool
+	excludeSource        bool
+	includeSourceTenancy bool
+	excludeSourceTenancy bool
+	destRules            []*destRules
+}
+
+type destRules struct {
+	values   *ruleValues
+	excludes []*ruleValues
+}
+
+type ruleValues struct {
+	portNames []string
+	path      string
+	pathPref  string
+	pathReg   string
+	headers   []string
+	methods   []string
+}
+
+type testResult struct {
+	fail    bool
+	port    string
+	path    string
+	headers map[string]string
+}
+
+func newTrafficPermissions(p *permission, srcTenancy *pbresource.Tenancy) *pbauth.TrafficPermissions {
```

Comment:
It's probably safer to manually construct traffic permissions rather than relying on a helper. It's definitely annoying, IMO being explicit will pay off with better readability.